Training/Program.cs

```diff
@@ -1,7 +1,76 @@
-﻿namespace Training {
-   internal class Program {
-      static void Main (string[] args) {
-         Console.WriteLine ("Hello, World!");
+﻿// --------------------------------------------------------------------------------------------
+// Training ~ A training program for new interns at Metamation, Batch - July 2023
+// Copyright (c) Metamation India.                                              
+// ------------------------------------------------------------------------
+// Program.cs
+// Implement file name parser with state machine. Every file name has the components drive letter A to Z, colon,
+// then one or more directory names followed by the actual file name which must always have an extension.
+// Only uppercase letters allowed just A to Z and all these components must be there and return the drive letter, 
+// the folder without any slash before or after and the actual file name. The return value can be a tuple with four strings.
+// --------------------------------------------------------------------------------------------
+using static EState;
+namespace Training {
+   #region Program ----------------------------------------------------------------------------
+   /// <summary>Implement file name parser with state machine</summary>
+   public class Program {
+      #region Methods -----------------------------------------------
+      /// <summary>Gets user input as a filepath and parse it</summary>
+      static void Main () {
+         for (; ; ) {
+            Console.Write ("File name: ");
+            string fileName = Console.ReadLine ();
+            if (fileName == "") break; // Exits when no input is provided.
+            try {
+               (string drive, string folder, string file, string ext) = GetParsedFilePath (fileName);
+               Console.Write ($"Drive     : {drive}\n" +
+                              $"Folder    : {folder}\n" +
+                              $"File      : {file}\n" +
+                              $"Extension : {ext}\n\n");
+            } catch (InValidPathException e) { Console.Write (e.Message + "\n\n"); }
+         }
       }
+
+      /// <summary>Parse the file path</summary>
+      /// <param name="filePath">String file path</param>
+      /// <returns>Returns drive, folder, filename and extension as tuple of four string</returns>
+      public static (string, string, string, string) GetParsedFilePath (string filePath) {
+         filePath = (filePath + '~').Replace ('/', '\\');
+         EState s = A;
+         int firstSlash = filePath.IndexOf ('\\'),
+             lastSlash = filePath.LastIndexOf ("\\"), dotIndex = filePath.IndexOf ('.');

```

Comment:
This will go through all the characters once to figure out  what is the index. But you going through all the characters again in line number 43. So, try to rewrite without using IndexOf.